main.py

```diff
@@ -14,9 +14,58 @@
 
 """URL routing definitions, and some basic error/warmup handlers."""
 
+#pylint: disable=bad-option-value,wrong-import-position,invalid-name
 from __future__ import absolute_import  # pylint: disable=import-only-modules
 from __future__ import unicode_literals  # pylint: disable=import-only-modules
 
+import os
+import sys
+
+# Root path of the app.
+ROOT_PATH = os.path.dirname(__file__)
+_PARENT_DIR = os.path.abspath(os.path.join(os.getcwd(), os.pardir))
+oppia_tools_path = os.path.join(_PARENT_DIR, 'oppia_tools')
+
+# oppia_tools/ is available locally (in both dev and prod mode). However,
+# on the GAE production server, oppia_tools/ is not available, and the default
+# PIL third-party library is used instead.
+#
+# We cannot special-case this using DEV_MODE because it is possible to run
+# Oppia in production mode locally, where a built-in PIL won't be available.
+# Hence the check for oppia_tools instead.
+if os.path.isdir(oppia_tools_path):
+    pil_path = os.path.join(
+        oppia_tools_path, 'Pillow-6.2.2')
+    if not os.path.isdir(pil_path):
+        raise Exception('Invalid path for oppia_tools library: %s' % pil_path)
+    sys.path.insert(0, pil_path)
+
+THIRD_PARTY_LIBS = [
+    os.path.join(
+        ROOT_PATH, 'third_party', 'backports.functools_lru_cache-1.6.1'),
+    os.path.join(ROOT_PATH, 'third_party', 'beautifulsoup4-4.9.1'),
+    os.path.join(ROOT_PATH, 'third_party', 'bleach-3.1.5'),
+    os.path.join(ROOT_PATH, 'third_party', 'callbacks-0.3.0'),
+    os.path.join(ROOT_PATH, 'third_party', 'future-0.17.1'),
+    os.path.join(ROOT_PATH, 'third_party', 'gae-cloud-storage-1.9.22.1'),
+    os.path.join(ROOT_PATH, 'third_party', 'gae-mapreduce-1.9.22.0'),
+    os.path.join(ROOT_PATH, 'third_party', 'gae-pipeline-1.9.22.1'),
+    os.path.join(ROOT_PATH, 'third_party', 'graphy-1.0.0'),
+    os.path.join(ROOT_PATH, 'third_party', 'html5lib-python-1.1'),
+    os.path.join(ROOT_PATH, 'third_party', 'mutagen-1.43.0'),
+    os.path.join(ROOT_PATH, 'third_party', 'packaging-20.4'),
+    os.path.join(ROOT_PATH, 'third_party', 'pylatexenc-2.6'),
+    os.path.join(ROOT_PATH, 'third_party', 'simplejson-3.17.0'),
+    os.path.join(ROOT_PATH, 'third_party', 'six-1.15.0'),
+    os.path.join(ROOT_PATH, 'third_party', 'soupsieve-1.9.5'),
+    os.path.join(ROOT_PATH, 'third_party', 'webencodings-0.5.1'),
+]
+
+for lib_path in THIRD_PARTY_LIBS:
+    if not os.path.isdir(lib_path):
+        raise Exception('Invalid path for third_party library: %s' % lib_path)
+    sys.path.insert(0, lib_path)
+
 import logging
```

Comment:
This is a system stdlib. Why does it need to be imported here, in a different location from os, sys above?

Also, if you decide to import things in weird places, there should be a very clear comment explaining the reason behind doing so and why doing it in other places doesn't work.